Handle % by zero errors

They should have proper error codes like divide by zero has
```
ionsql> 1 % 0
      |
com.amazon.ionsql.eval.EvaluationException: Internal error, / by zero
	<UNKNOWN>: at line 1, column 3: <UNKNOWN>

	at com.amazon.ionsql.eval.RuntimeExprThunk.eval(EvaluatingCompiler.kt:51)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compile$2.eval(EvaluatingCompiler.kt:1548)
	at com.amazon.ionsql.eval.EvaluatingCompiler.eval(EvaluatingCompiler.kt:1527)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compile$3.eval(EvaluatingCompiler.kt:1560)
	at com.amazon.ionsql.tools.REPL.main(REPL.kt:186)
Caused by: java.lang.ArithmeticException: / by zero
	at com.amazon.ionsql.util.NumberExtensionsKt.rem(NumberExtensions.kt:199)
	at com.amazon.ionsql.eval.EvaluatingCompiler$syntaxTable$16.invoke(EvaluatingCompiler.kt:456)
	at com.amazon.ionsql.eval.EvaluatingCompiler$syntaxTable$16.invoke(EvaluatingCompiler.kt:101)
	at com.amazon.ionsql.eval.ExprFunction$Companion$over$1.call(ExprFunction.kt:19)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compileFunc$2.invoke(EvaluatingCompiler.kt:1472)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compileFunc$2.invoke(EvaluatingCompiler.kt:101)
	at com.amazon.ionsql.eval.EvaluatingCompiler$exprThunk$1.innerEval(EvaluatingCompiler.kt:120)
	at com.amazon.ionsql.eval.RuntimeExprThunk.eval(EvaluatingCompiler.kt:33)
	... 4 more

ERROR! (49 ms)
ionsql> 1 / 0
      |
com.amazon.ionsql.eval.EvaluationException: / by zero
	<UNKNOWN>: at line 1, column 3: <UNKNOWN>

	at com.amazon.ionsql.eval.RuntimeExprThunk.eval(EvaluatingCompiler.kt:37)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compile$2.eval(EvaluatingCompiler.kt:1548)
	at com.amazon.ionsql.eval.EvaluatingCompiler.eval(EvaluatingCompiler.kt:1527)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compile$3.eval(EvaluatingCompiler.kt:1560)
	at com.amazon.ionsql.tools.REPL.main(REPL.kt:186)
Caused by: com.amazon.ionsql.eval.EvaluationException: / by zero
	<UNKNOWN>: at line <UNKNOWN>, column <UNKNOWN>: <UNKNOWN>

	at com.amazon.ionsql.eval.ExceptionsKt.err(Exceptions.kt:44)
	at com.amazon.ionsql.eval.ExceptionsKt.errNoContext(Exceptions.kt:40)
	at com.amazon.ionsql.eval.EvaluatingCompiler$syntaxTable$15.invoke(EvaluatingCompiler.kt:448)
	at com.amazon.ionsql.eval.EvaluatingCompiler$syntaxTable$15.invoke(EvaluatingCompiler.kt:101)
	at com.amazon.ionsql.eval.ExprFunction$Companion$over$1.call(ExprFunction.kt:19)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compileFunc$2.invoke(EvaluatingCompiler.kt:1472)
	at com.amazon.ionsql.eval.EvaluatingCompiler$compileFunc$2.invoke(EvaluatingCompiler.kt:101)
	at com.amazon.ionsql.eval.EvaluatingCompiler$exprThunk$1.innerEval(EvaluatingCompiler.kt:120)
	at com.amazon.ionsql.eval.RuntimeExprThunk.eval(EvaluatingCompiler.kt:33)
	... 4 more
```

--- a/lang/src/org/partiql/lang/errors/ErrorCode.kt
+++ b/lang/src/org/partiql/lang/errors/ErrorCode.kt
@@ -552,6 +552,16 @@ enum class ErrorCode(private val category: ErrorCategory,
         LOCATION,
         "LIMIT must not be negative"),
 
+    EVALUATOR_DIVIDE_BY_ZERO(
+        ErrorCategory.EVALUATOR,
+        LOCATION,
+        "/ by zero"),
+
+    EVALUATOR_MODULO_BY_ZERO(
+        ErrorCategory.EVALUATOR,
+        LOCATION,
+        "% by zero"),
+
     SEMANTIC_ILLEGAL_GLOBAL_VARIABLE_ACCESS(
         ErrorCategory.SEMANTIC,
         LOCATION + setOf(Property.BINDING_NAME),

--- a/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt
+++ b/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt
@@ -421,7 +421,7 @@ internal class EvaluatingCompiler(
         return thunkFold(valueFactory.nullValue, metas, argThunks) { lValue, rValue ->
             val denominator = rValue.numberValue()
             if (denominator.isZero()) {
-                errNoContext("/ by zero", internal = false)
+                err("/ by zero", ErrorCode.EVALUATOR_DIVIDE_BY_ZERO, null, false)
             }
             try {
                 (lValue.numberValue() / denominator).exprValue()
@@ -438,7 +438,14 @@ internal class EvaluatingCompiler(
         return thunkFold(
             valueFactory.nullValue,
             metas,
-            argThunks) { lValue, rValue -> (lValue.numberValue() % rValue.numberValue()).exprValue() }
+            argThunks) { lValue, rValue ->
+            val denominator = rValue.numberValue()
+            if (denominator.isZero()) {
+                err("% by zero", ErrorCode.EVALUATOR_MODULO_BY_ZERO, null, false)
+            }
+            
+            (lValue.numberValue() % denominator).exprValue() 
+        }
     }
 
     private fun compileNAryEq(

--- a/lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt
+++ b/lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt
@@ -142,14 +142,28 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
         IllegalArgumentException::class)
 
     @Test
-    fun divideByZero() = assertThrows("/ by zero", NodeMetadata(1, 3)) {
-        voidEval("1 / 0")
-    }
+    fun divideByZero() = checkInputThrowingEvaluationException(
+        "1 / 0",
+        ErrorCode.EVALUATOR_DIVIDE_BY_ZERO,
+        sourceLocationProperties(1, 3)) 
 
     @Test
-    fun divideByZeroDecimal() = assertThrows("/ by zero", NodeMetadata(1, 5)) {
-        voidEval("1.0 / 0.0")
-    }
+    fun divideByZeroDecimal() = checkInputThrowingEvaluationException(
+        "1.0 / 0.0",
+        ErrorCode.EVALUATOR_DIVIDE_BY_ZERO,
+        sourceLocationProperties(1, 5))
+
+    @Test
+    fun moduloByZero() = checkInputThrowingEvaluationException(
+        "1 % 0",
+        ErrorCode.EVALUATOR_MODULO_BY_ZERO,
+        sourceLocationProperties(1, 3))
+
+    @Test
+    fun moduloByZeroDecimal() = checkInputThrowingEvaluationException(
+        "1.0 % 0.0",
+        ErrorCode.EVALUATOR_MODULO_BY_ZERO,
+        sourceLocationProperties(1, 5))
 
     @Test
     fun divideByZeroInSelect() = assertThrows("/ by zero", NodeMetadata(1, 76)) {

